Handling plugin exit code

By default, [pytest's exit code](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.ExitCode) follows the usual convention of returning 0 on success (no test failures) and non-zero values otherwise. It might be nice for the plugin to set the exit code to 0 if the only failures in a test suite run are flaky failures. This behavior could be toggled via a command line option. The [`pytest_sessionfinish`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_sessionfinish) hook is likely to be helpful here.

What exactly does the exit code do here? #8 implements this, but it still prints the "failures" out:
```
================================================ FAILURES ================================================
____________________________________________ test_eqiulateral ____________________________________________

    def test_eqiulateral():
>       assert triangle_type(1, 1, 1) == "Equilateral"
E       AssertionError: assert 'Equilateral Triangle' == 'Equilateral'
E         
E         - Equilateral
E         + Equilateral Triangle

triangle.py:11: AssertionError

```
Do we want to surpress this output?

I also can't see any mention of the exit code on the output. Is this expected?

--- a/.github/workflows/ci-tests-drafts.yaml
+++ b/.github/workflows/ci-tests-drafts.yaml
@@ -29,7 +29,7 @@ jobs:
           pip install pytest pytest-cov
     - name: Test with pytest
       run: |
-        pytest -p no:flakefighter --runpytest=subprocess --cov=src --cov-report=xml
+        pytest -p no:flakefighter --cov=src --cov-report=xml
     - name: "Upload coverage to Codecov"
       uses: codecov/codecov-action@v5
       with:

--- a/.github/workflows/ci-tests.yaml
+++ b/.github/workflows/ci-tests.yaml
@@ -34,7 +34,7 @@ jobs:
           pip install pytest pytest-cov
     - name: Test with pytest
       run: |
-        pytest -p no:flakefighter --runpytest=subprocess --cov=src --cov-report=xml
+        pytest -p no:flakefighter --cov=src --cov-report=xml
     - name: "Upload coverage to Codecov"
       uses: codecov/codecov-action@v5
       with:

--- a/src/pytest_flakefighter/plugin.py
+++ b/src/pytest_flakefighter/plugin.py
@@ -80,7 +80,7 @@ class FlakeFighter:
             ):
                 report.flaky = True
             else:
-                self.genunine_failures_observed = True
+                self.genuine_failure_observed = True
         return report
 
     def pytest_report_teststatus(
@@ -94,7 +94,6 @@ class FlakeFighter:
         """
         if getattr(report, "flaky", False):
             return report.outcome, "F", ("FLAKY", {"yellow": True})
-        self.genuine_failure_observed = True
         return None
 
     def line_modified_by_latest_commit(self, file_path: str, line_number: int) -> bool:
@@ -119,6 +118,7 @@ class FlakeFighter:
         :param session: The pytest session object.
         :param exitstatus: The status which pytest will return to the system.
         """
+        print("\nSTATUS", session.config.option.suppress_flaky, not self.genuine_failure_observed)
         if session.config.option.suppress_flaky and not self.genuine_failure_observed:
             session.exitstatus = pytest.ExitCode.OK
 

--- a/src/pytest_flakefighter/plugin.py
+++ b/src/pytest_flakefighter/plugin.py
@@ -25,6 +25,7 @@ class FlakeFighter:
         self.lines_changed = {
             os.path.abspath(os.path.join(root, file)): {} for file in self.repo.commit(self.commit).stats.files
         }
+        self.genuine_failure_observed = False
 
     @pytest.hookimpl(hookwrapper=True)
     def pytest_runtest_protocol(self, item: pytest.Item, nextitem: pytest.Item):  # pylint: disable=unused-argument
@@ -64,6 +65,8 @@ class FlakeFighter:
                 if file_path in self.lines_changed
             ):
                 report.flaky = True
+            else:
+                self.genunine_failures_observed = True
         return report
 
     def pytest_report_teststatus(
@@ -77,6 +80,7 @@ class FlakeFighter:
         """
         if getattr(report, "flaky", False):
             return report.outcome, "F", ("FLAKY", {"yellow": True})
+        self.genuine_failure_observed = True
         return None
 
     def line_modified_by_latest_commit(self, file_path: str, line_number: int) -> bool:
@@ -91,6 +95,18 @@ class FlakeFighter:
         output = self.repo.git.log("-L", f"{line_number},{line_number}:{file_path}")
         self.lines_changed[file_path][line_number] = f"commit {self.commit}" in output
         return self.lines_changed[file_path][line_number]
+
+    def pytest_sessionfinish(
+        self,
+        session: pytest.Session,
+        exitstatus: pytest.ExitCode,  # pylint: disable=unused-argument
+    ) -> None:
+        """Called after whole test run finished, right before returning the exit status to the system.
+        :param session: The pytest session object.
+        :param exitstatus: The status which pytest will return to the system.
+        """
+        if session.config.option.suppress_flaky and not self.genuine_failure_observed:
+            session.exitstatus = pytest.ExitCode.OK
 
 
 def pytest_addoption(parser: pytest.Parser):
@@ -112,6 +128,13 @@ def pytest_addoption(parser: pytest.Parser):
         dest="repo_path",
         default=None,
         help="The commit hash to compare against.",
+    )
+    group.addoption(
+        "--suppress-flaky-failures-exit-code",
+        action="store_true",
+        dest="suppress_flaky",
+        default=False,
+        help="Return OK exit code if the only failures are flaky failures.",
     )
 
 

--- a/src/pytest_flakefighter/plugin.py
+++ b/src/pytest_flakefighter/plugin.py
@@ -27,15 +27,31 @@ class FlakeFighter:
         }
         self.genuine_failure_observed = False
 
-    @pytest.hookimpl(hookwrapper=True)
-    def pytest_runtest_protocol(self, item: pytest.Item, nextitem: pytest.Item):  # pylint: disable=unused-argument
+    def pytest_sessionstart(self, session: pytest.Session):  # pylint: disable=unused-argument
         """
-        Perform the runtest protocol for a single test item.
-
-        :param item: Test item for which the runtest protocol is performed.
-        :param nextitem: The scheduled-to-be-next test item (or None if this is the end my friend).
+        Start the coverage measurement before tests are collected so we measure class and method definitions as covered.
+        :param session: The session.
         """
         self.cov.start()
+
+    def pytest_collection_finish(self, session: pytest.Session):  # pylint: disable=unused-argument
+        """
+        Stop the coverage measurement after tests are collected.
+        :param session: The session.
+        """
+        # This line cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
+        self.cov.stop()
+
+    @pytest.hookimpl(hookwrapper=True)
+    def pytest_runtest_call(self, item: pytest.Item):
+        """
+        Start the coverage measurement and label the coverage for the current test, run the test,
+        then stop coverage measurement.
+
+        :param item: The item.
+        """
+        self.cov.start()
+        # These lines cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
         self.cov.switch_context(item.nodeid)
         yield
         self.cov.stop()
@@ -45,8 +61,7 @@ class FlakeFighter:
         self, item: pytest.Item, call: pytest.CallInfo[None]  # pylint: disable=unused-argument
     ) -> pytest.TestReport:
         """
-        Called to create a :class:`~pytest.TestReport` for each of
-        the setup, call and teardown runtest phases of a test item.
+        Classify failed tests as flaky if they don't cover any changed code.
 
         :param item: The item.
         :param call: The :class:`~pytest.CallInfo` for the phase.
@@ -57,7 +72,6 @@ class FlakeFighter:
         report = outcome.get_result()
         if report.when == "call" and report.failed:
             line_coverage = self.cov.get_data()
-            line_coverage.set_query_context(item.nodeid)
             if not any(
                 self.line_modified_by_latest_commit(file_path, line_number)
                 for file_path in line_coverage.measured_files()
@@ -73,7 +87,7 @@ class FlakeFighter:
         self, report: pytest.TestReport, config: pytest.Config  # pylint: disable=unused-argument
     ) -> tuple[str, str, str]:
         """
-        Return result-category, shortletter and verbose word for status reporting.
+        Report flaky failures as such.
         :param report: The report object whose status is to be returned.
         :param config: The pytest config object.
         :returns: The test status.

--- a/src/pytest_flakefighter/plugin.py
+++ b/src/pytest_flakefighter/plugin.py
@@ -119,7 +119,7 @@ class FlakeFighter:
         :param exitstatus: The status which pytest will return to the system.
         """
         print("\nSTATUS", session.config.option.suppress_flaky, not self.genuine_failure_observed)
-        if session.config.option.suppress_flaky and not self.genuine_failure_observed:
+        if session.config.option.suppress_flaky and session.exitstatus == pytest.ExitCode.TESTS_FAILED and not self.genuine_failure_observed:
             session.exitstatus = pytest.ExitCode.OK
 
 

--- a/src/pytest_flakefighter/plugin.py
+++ b/src/pytest_flakefighter/plugin.py
@@ -39,7 +39,7 @@ class FlakeFighter:
         Stop the coverage measurement after tests are collected.
         :param session: The session.
         """
-        # This line cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
+        # Line cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
         self.cov.stop()
 
     @pytest.hookimpl(hookwrapper=True)
@@ -51,7 +51,7 @@ class FlakeFighter:
         :param item: The item.
         """
         self.cov.start()
-        # These lines cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
+        # Lines cannot appear as covered on our tests because the coverage measurement is leaking into the self.cov
         self.cov.switch_context(item.nodeid)
         yield
         self.cov.stop()

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -12,13 +12,13 @@ import pytest
 pytest_plugins = "pytester"
 
 
-@pytest.fixture(scope="session", name="triangle_repo")
-def fixture_triangle_repo(tmpdir_factory):
+@pytest.fixture(scope="session", name="flaky_triangle_repo")
+def fixture_flaky_triangle_repo(tmpdir_factory):
     """
-    Fixture for a minimal git repo with a commit history of broken tests.
+    Fixture for a minimal git repo with a commit history to hide failing tests.
     """
-    repo_root = tmpdir_factory.mktemp("triangle_repo")
-    repo = git.Repo.init(repo_root)
+    repo_root = tmpdir_factory.mktemp("flaky_triangle_repo")
+    repo = git.Repo.init(repo_root, initial_branch="main")
     shutil.copy(os.path.join("tests", "resources", "triangle_example.txt"), os.path.join(repo_root, "triangle.py"))
     repo.index.add(["triangle.py"])
     repo.index.commit("Initial commit of test file.")
@@ -26,6 +26,8 @@ def fixture_triangle_repo(tmpdir_factory):
     shutil.copy(os.path.join("tests", "resources", "triangle_broken.txt"), os.path.join(repo_root, "triangle.py"))
     repo.index.add(["triangle.py"])
     repo.index.commit("Broke the tests.")
+
+    repo.index.commit("This is an empty commit")
     return repo_root
 
 
@@ -35,7 +37,7 @@ def fixture_deflaker_repo(tmpdir_factory):
     Fixture for a minimal git repo with a commit history of broken tests.
     """
     repo_root = tmpdir_factory.mktemp("deflaker_repo")
-    repo = git.Repo.init(repo_root)
+    repo = git.Repo.init(repo_root, initial_branch="main")
     shutil.copy(os.path.join("tests", "resources", "deflaker_example.txt"), os.path.join(repo_root, "app.py"))
     repo.index.add(["app.py"])
     repo.index.commit("Initial commit of test file.")

--- a/tests/test_flakefighter.py
+++ b/tests/test_flakefighter.py
@@ -4,44 +4,68 @@ Test DeFlaker algorithm.
 
 import os
 
+import git
 
-def test_files_exist(triangle_repo, deflaker_repo):
+
+def test_files_exist(flaky_triangle_repo, deflaker_repo):
     """
     Test that the necessary fixture files have actually heen created.
     """
     assert os.path.exists(
-        os.path.join(triangle_repo, "triangle.py")
-    ), f"Fixture file {os.path.join(triangle_repo, 'triangle.py')} not present."
+        os.path.join(flaky_triangle_repo, "triangle.py")
+    ), f"Fixture file {os.path.join(flaky_triangle_repo, 'triangle.py')} not present."
     assert os.path.exists(
         os.path.join(deflaker_repo, "app.py")
     ), f"Fixture file {os.path.join(deflaker_repo, 'app.py')} not present."
 
 
-def test_real_failures(pytester, triangle_repo):
-    """Make sure that pytest accepts our fixture."""
+def test_real_failures(pytester, flaky_triangle_repo):
+    """Make sure that genuine failures are labelled as such."""
+    repo = git.Repo(flaky_triangle_repo)
+    commits = [commit.hexsha for commit in repo.iter_commits("main")]
 
-    # run pytest with the following cmd args
-    result = pytester.runpytest(os.path.join(triangle_repo, "triangle.py"), f"--repo={triangle_repo}", "-s")
+    result = pytester.runpytest(
+        os.path.join(flaky_triangle_repo, "triangle.py"),
+        f"--repo={flaky_triangle_repo}",
+        f"--commit={commits[1]}",
+        "-s",
+    )
 
     result.assert_outcomes(failed=3)
     result.stdout.fnmatch_lines(
         [
-            f"FAILED {os.path.join('..','triangle_repo0', 'triangle.py')}::test_eqiulateral*",
-            f"FAILED {os.path.join('..','triangle_repo0', 'triangle.py')}::test_isosceles*",
-            f"FAILED {os.path.join('..','triangle_repo0', 'triangle.py')}::test_scalene*",
+            f"FAILED {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_eqiulateral*",
+            f"FAILED {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_isosceles*",
+            f"FAILED {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_scalene*",
         ]
     )
 
 
-# def test_deflaker_example(pytester, deflaker_repo):
-#     """Make sure that pytest accepts our fixture."""
-#
-#     # run pytest with the following cmd args
-#     result = pytester.runpytest(
-#         os.path.join(deflaker_repo, "app.py"),
-#         f"--repo={deflaker_repo}",
-#     )
-#
-#     assert result.ret == 1, "Expected tests to fail"
-#
-#     result.stdout.fnmatch_lines(["Real faults ['deflaker_repo0/app.py::test_app']"])
+def test_flaky_failures(pytester, flaky_triangle_repo):
+    """Make sure that flaky failures are labelled as such"""
+
+    # run pytest with the following cmd args
+    result = pytester.runpytest(os.path.join(flaky_triangle_repo, "triangle.py"), f"--repo={flaky_triangle_repo}", "-s")
+
+    result.assert_outcomes(failed=3)
+    result.stdout.fnmatch_lines(
+        [
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_eqiulateral*",
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_isosceles*",
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_scalene*",
+        ]
+    )
+
+
+def test_deflaker_example(pytester, deflaker_repo):
+    """Make sure that pytest accepts our fixture."""
+
+    # run pytest with the following cmd args
+    result = pytester.runpytest(
+        os.path.join(deflaker_repo, "app.py"),
+        f"--repo={deflaker_repo}",
+    )
+
+    assert result.ret == 1, "Expected tests to fail"
+
+    result.stdout.fnmatch_lines([f"FAILED {os.path.join('..','deflaker_repo0', 'app.py')}::test_app - assert False"])

--- a/tests/test_flakefighter.py
+++ b/tests/test_flakefighter.py
@@ -5,6 +5,7 @@ Test DeFlaker algorithm.
 import os
 
 import git
+from pytest import ExitCode
 
 
 def test_files_exist(flaky_triangle_repo, deflaker_repo):
@@ -39,13 +40,18 @@ def test_real_failures(pytester, flaky_triangle_repo):
             f"FAILED {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_scalene*",
         ]
     )
+    assert result.ret == ExitCode.TESTS_FAILED, f"Expected exit code {ExitCode.TESTS_FAILED} but was {result.ret}."
 
 
 def test_flaky_failures(pytester, flaky_triangle_repo):
     """Make sure that flaky failures are labelled as such"""
 
     # run pytest with the following cmd args
-    result = pytester.runpytest(os.path.join(flaky_triangle_repo, "triangle.py"), f"--repo={flaky_triangle_repo}", "-s")
+    result = pytester.runpytest(
+        os.path.join(flaky_triangle_repo, "triangle.py"),
+        f"--repo={flaky_triangle_repo}",
+        "-s",
+    )
 
     result.assert_outcomes(failed=3)
     result.stdout.fnmatch_lines(
@@ -55,6 +61,29 @@ def test_flaky_failures(pytester, flaky_triangle_repo):
             f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_scalene*",
         ]
     )
+    assert result.ret == ExitCode.TESTS_FAILED, f"Expected exit code {ExitCode.TESTS_FAILED} but was {result.ret}."
+
+
+def test_suppress_flaky_failures(pytester, flaky_triangle_repo):
+    """Make sure that flaky failures are labelled as such"""
+
+    # run pytest with the following cmd args
+    result = pytester.runpytest(
+        os.path.join(flaky_triangle_repo, "triangle.py"),
+        f"--repo={flaky_triangle_repo}",
+        "--suppress-flaky-failures-exit-code",
+        "-s",
+    )
+
+    result.assert_outcomes(failed=3)
+    result.stdout.fnmatch_lines(
+        [
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_eqiulateral*",
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_isosceles*",
+            f"FLAKY {os.path.join('..','flaky_triangle_repo0', 'triangle.py')}::test_scalene*",
+        ]
+    )
+    assert result.ret == ExitCode.OK, f"Expected exit code {ExitCode.OK} but was {result.ret}."
 
 
 def test_deflaker_example(pytester, deflaker_repo):
@@ -66,6 +95,6 @@ def test_deflaker_example(pytester, deflaker_repo):
         f"--repo={deflaker_repo}",
     )
 
-    assert result.ret == 1, "Expected tests to fail"
-
+    result.assert_outcomes(failed=1)
     result.stdout.fnmatch_lines([f"FAILED {os.path.join('..','deflaker_repo0', 'app.py')}::test_app - assert False"])
+    assert result.ret == ExitCode.TESTS_FAILED, f"Expected exit code {ExitCode.TESTS_FAILED} but was {result.ret}."

